### PR TITLE
fix(core): non-collapsed feed fields content

### PR DESCRIPTION
**Pull request by Rongronggg9**
## 该 PR 相关 Issue / Involved issue

#9259

## 完整路由地址 / Example for the proposed route(s)

```routes
NOROUTE
```

FYI, the route in the screenshots below is `/amazon/ku/this`.

## 新 RSS 检查列表 / New RSS Script Checklist
  
- [ ] 新的路由 New Route
  - [ ] 跟随 [v2 路由规范](https://docs.rsshub.app/joinus/script-standard.html) Follows [v2 Script Standard](https://docs.rsshub.app/en/joinus/script-standard.html)
- [ ] 文档说明 Documentation
  - [ ] 中文文档 CN
  - [ ] 英文文档 EN
- [ ] 全文获取 fulltext
  - [ ] 使用缓存 Use Cache
- [ ] 反爬/频率限制 anti-bot or rate limit?
  - [ ] 如果有, 是否有对应的措施? If yes, do your code reflect this sign?
- [ ] [日期和时间](https://docs.rsshub.app/joinus/pub-date.html) [date and time](https://docs.rsshub.app/en/joinus/pub-date.html)
  - [ ] 可以解析 Parsed
  - [ ] 时区调整 Correct TimeZone
- [ ] 添加了新的包 New package added
- [ ] `Puppeteer`

## 说明 / Note
Unintended linebreaks, white spaces in some fields will be trimmed ([ref](https://developer.mozilla.org/en-US/docs/Web/CSS/white-space)) by most RSS readers while rendering a post, but can break some RSS readers:
```xml
            <author><![CDATA[
      冯骥才       
       (作者) 
        ]]></author>
```

Thus, trimming them (just like `white-space: normal;`, but the spaces at the start/end are trimmed additionally) does not harm but can avoid many problems:
```xml
            <author><![CDATA[冯骥才 (作者)]]></author>
```

**What makes things worse is that if the whitespaces and linebreaks in the entry title uncollapsed, RSSHub will count those whitespaces and linebreaks during trimming the post title, making the post title even shorter than intended.**

### CODE REVIEW

File: lib/middleware/template.js [hunk 1]
```diff
@@ -40,10 +40,22 @@ module.exports = async (ctx, next) => {
...
+                ctx.state.data.title = ctx.state.data.title.replace(/\s+/g, ' ').trim();
+            }
+
+            if (ctx.state.data.subtitle) {
+                ctx.state.data.subtitle = ctx.state.data.subtitle.replace(/\s+/g, ' ').trim();
+            }
+
+            if (ctx.state.data.author) {
+                ctx.state.data.author = ctx.state.data.author.replace(/\s+/g, ' ').trim();
+            }
```
Comment (SukkaW) [inline lib/middleware/template.js:53]:
```suggestion
            const rWhiteSpace = /\s+/;
            const rAllWhiteSpace = /\s+/g;
            if (ctx.state.data.title && rWhiteSpace.test(ctx.state.data.title)) {
                ctx.state.data.title = ctx.state.data.title.replace(rAllWhiteSpace, ' ').trim();
            }
            if (ctx.state.data.subtitle && rWhiteSpace.test(ctx.state.data.subtitle)) {
                ctx.state.data.subtitle = ctx.state.data.subtitle.replace(rAllWhiteSpace, ' ').trim();
            }
            if (ctx.state.data.author && rWhiteSpace.test(ctx.state.data.author)) {
                ctx.state.data.author = ctx.state.data.author.replace(rAllWhiteSpace, ' ').trim();
            }
```

File: lib/middleware/template.js [hunk 2]
```diff
@@ -40,10 +40,22 @@ module.exports = async (ctx, next) => {
...
+
             ctx.state.data.item &&
                 ctx.state.data.item.forEach((item) => {
                     if (item.title) {
```
Comment (SukkaW) [inline lib/middleware/template.js:57]:
```suggestion
                    if (item.title && rWhiteSpace.test(item.title)) {
```

File: lib/middleware/template.js [hunk 3]
```diff
@@ -40,10 +40,22 @@ module.exports = async (ctx, next) => {
...
-                        item.title = item.title.trim();
+                        item.title = item.title.replace(/\s+/g, ' ').trim();
```
Comment (SukkaW) [inline lib/middleware/template.js:58]:
```suggestion
                        item.title = item.title.replace(rAllWhiteSpace, ' ').trim();
```

File: lib/middleware/template.js [hunk 4]
```diff
@@ -54,6 +66,10 @@ module.exports = async (ctx, next) => {
                         }
                     }
 
+                    if (item.author) {
+                        item.author = item.author.replace(/\s+/g, ' ').trim();
+                    }
```
Comment (SukkaW) [inline lib/middleware/template.js:71]:
```suggestion
                    if (item.author && rWhiteSpace.test(item.author)) {
                        item.author = item.author.replace(rAllWhiteSpace, ' ').trim();
                    }
```

**Review by SukkaW — CHANGES_REQUESTED** [review]
This part of the middleware will be a hot path, so we'd better make sure it runs fast and not add too many overheads.

Besides my current suggestions, there are also more that can be done, e.g. by extracting a `collapseWhitespace` function, it enables v8 to optimize the call and further utilize the v8 inline cache.

**Comment by Rongronggg9** [comment]
> extracting a `collapseWhitespace` function, it enables v8 to optimize the call and further utilize the v8 inline cache.

OK, I've done this. Please check if I made something wrong or not, since I am not so familiar with these speedup techniques.

File: lib/middleware/template.js (hunk 2 above)
Comment (Rongronggg9) [reply lib/middleware/template.js:50]: We should not do this, since a title trimming must happen even if there is no whitespace to be collapsed.

**Comment by Rongronggg9** [comment]
> What makes things worse is that if the whitespaces and linebreaks in the entry title uncollapsed, RSSHub will count those whitespaces and linebreaks during trimming the post title, making the post title even shorter than intended.

Here is a severe example:
https://rsshub.app/newzmz
```
            <title><![CDATA[第四季合集                                
                                    了不起的麦瑟尔夫人 第四季
                                    The Marve...]]></title>
```
There are 32 whitespaces and 1 linebreak after `第四季合集`, 36 whitespaces before and 1 linebreak after `了不起的麦瑟尔夫人 第四季`, and 36 whitespaces before `The Marvelous Mrs. Maisel`, causing unintended title trimming severely.

**Review by SukkaW — APPROVED** [review]

**Pull request merged**